hummingbot/connector/budget_checker.py

```diff
@@ -101,130 +209,169 @@ def adjust_candidate(
+        return order_candidate
 
-    def _get_collateral_token(self, order_candidate: OrderCandidate) -> str:
+    def _get_order_collateral_token(self, order_candidate: OrderCandidate) -> Optional[str]:
         trading_pair = order_candidate.trading_pair
         base, quote = split_hb_trading_pair(trading_pair)
         if order_candidate.order_side == TradeType.BUY:
-            collateral_token = quote
+            oc_token = quote
         else:
-            collateral_token = base
-        return collateral_token
+            oc_token = base
+        return oc_token
 
-    def _get_base_required_collateral(self, order_candidate: OrderCandidate) -> Decimal:
+    def _get_order_collateral_amount(
+        self, order_candidate: OrderCandidate, order_collateral_token: str
+    ) -> Decimal:
         order_size, size_token = self._get_order_size_and_size_token(order_candidate)
-        size_collateral_price = self._get_size_collateral_price(order_candidate)
-        required_collateral = order_size * size_collateral_price
-        return required_collateral
-
-    def _reduce_order_and_collateral_amounts(
-        self, order_candidate: OrderCandidate, available_collateral: Decimal
-    ) -> OrderCandidate:
-        adjusted_amount = self._get_adjusted_amount(order_candidate, available_collateral)
-        adjusted_amount = self._quantize_adjusted_amount(order_candidate, adjusted_amount)
-
-        adjusted_candidate = copy(order_candidate)
-        adjusted_candidate.amount = adjusted_amount
-        adjusted_candidate.collateral_amount = available_collateral
-
-        return adjusted_candidate
+        size_collateral_price = self._get_size_collateral_price(order_candidate, order_collateral_token)
+        oc_amount = order_size * size_collateral_price
+        return oc_amount
+
+    def _populate_percent_fee_collateral_entry(self, order_candidate: OrderCandidate, fee: TradeFee) -> OrderCandidate:
+        if fee.percent_token is None:
+            fee.percent_token = order_candidate.order_collateral.token
+        if fee.percentage_application == TradeFeePercentageApplication.AddedToCost:
+            if order_candidate.order_collateral is None or fee.percent_token != order_candidate.order_collateral.token:
+                size, token = self._get_order_size_and_size_token(order_candidate)
+                if fee.percent_token == token:
+                    exchange_rate = Decimal("1")
+                else:
+                    exchange_pair = combine_to_hb_trading_pair(fee.percent_token, token)
+                    exchange_rate = self._exchange.get_price(exchange_pair, is_buy=True)
+                amount = size * exchange_rate * fee.percent
+            else:  # fee.percent_token == order_candidate.order_collateral.token
+                amount = order_candidate.order_collateral.amount * fee.percent
+            order_candidate.percent_fee_collateral = SizeEntry(fee.percent_token, amount)
```

Comment:
This decision made based on the fee application type makes me think this logic should be in an object instead.

Reply:
I agree that this is messy, and is also the main reason I wanted to open the draft PR for review; maybe we can discuss this together over a call.

My worry with pushing this into the the `TradeFee` itself is that it requires access to the exchange, and the essence of the logic is budget-related (using the trade fee). Both of those are higher level than the `TradeFee` object is intended to be.

The reason I went with this approach is because I don't think there will be a need for other branches in this if-else tree:
- If the fee is applied on the returns of the trade, then we're done.
- If the percent token and the order collateral tokens are the same, then we simply multiply the collateral amount by the percent and we're done.
- Otherwise, we get the size of the order and the token that it is sized in, the conversion rate between the size token and the fee token, and finally we derive the fee amount based on that information.

I don't see another case scenario here.

All that being said, I'd be curious to discus the possibility to put that logic in an object that wraps the `TradeFee` (?)

Reply:
> This decision made based on the fee application type makes me think this logic should be in an object instead.

@CoinAlpha/qa-team, after requesting your review, I realized I forgot to address this comment with @aarmoa. Please hold off on testing until he and I discuss it further as this may require some further code changes.

Reply:
In my last two commits below I have pushed the logic from the `BudgetChecker` into the respective `OrderCandidate` and the new `TradeFee` objects, addressing this issue.